gcs_light_client.mediawiki

```diff
@@ -0,0 +1,1056 @@
+server. Additionally, the compact client-side chain index also opens up the
+door to smart contract applications which require agent action in response to
+on-chain events. Examples of such applications include Lightning.
+
+== Backwards Compatability ==
+
+This light client protocl is NOT backwards compatible with BIP 37. Full nodes
+MAY implement both protocols to serve both types of light clients.
+
+
+== Implementation Notes ==
+
+This filter header chain SHOULD be utilized by light clients to gain a greater
+degree of security against bamboozling full-nodes during their initial chain
+sync. In addition to fetching all the bitcoin headers, light clients
+implementing this BIP SHOULD also fetch all the ''filter headers'' from
+''each'' of their connected peers. With these headers, light clients SHOULD
+efficiently detect nodes that advertise a conflicting filter chain history. To
+do this, light clients MUST ensure that all nodes return the ``same` filter
+header hash for a particular block header hash. 
+
+Light clients MAY use the filter header chain to verify purported filter
+authenticity when fetching the next set of headers from chain tip. Light
+clients MAY use the following algorithm to more efficiently verify the
+authenticity of filters (the naive version would fetch the entire filter from
+each peer, this version saves bandwidth):
+
+<pre>
+verify_from_tip(tip_block_hash: [32]byte):
+    let filter_types = {supported_filter_types...}
+    let connected_peers = {list_of_connected_full_nodes...}
+
+    for filter_type in filter_types:
+
+        let filter_headers = set()
+        for peer in connected_peers:
+            let filter_header = peer.fetch_filter_header(tip_block_hash)
+            filter_headers.insert(filter_header)
+
+        if len(filter_headers) != 1:
+            // Peers have conflicting filters. The light client should fetch
+            // each unique filter from the set of peers AND fetch the block. The
+            // light client can then verify which filter header is correct, and
+            // BAN the offending peers.
+
+        // Otherwise, syncing continues as normal: fetch filter to see if it
+        // matches any relevant items.
+</pre>
+
+Light clients MAY persistently commit all filter headers to disk, as when
+lazily fetching filters (due to a historical re-scan or chain analysis),
+they're able to verify the authenticity of any fetched filters.
+
+Full-nodes MAY persistently compute and persist the filter header chain
+on-disk, just as the regular filters.
+
+Zero-length filters are sent without an <code>N</code> value, allowing us to save
+<code>4-bytes</code>. Clients are able to verify that a filter will be
+<code>null</code> before requesting it (as it will just be the prior filter
+header hashed with zero bytes). Clients can take this fact into account in
+order to save a round trip for <code>null</code> headers.
+
+Light clients implementing this proposal SHOULD utilize the
```

Comment:
This seems very strong. Why SHOULD? Hell, why even MAY? It seems very unrelated.